File: generators/lines.cpp

// Part of the ProceduralTextureMaker project.
// http://github.com/johanokl/ProceduralTextureMaker
// Released under GPLv3.
// Johan Lindqvist (johan.lindqvist@gmail.com)

#include "lines.h"
#include <QColor>
#include <QtMath>
#include <cmath>

LinesTextureGenerator::LinesTextureGenerator() {
   TextureGeneratorSetting color;
   color.name = "Line colour";
   color.description = "Colour used to draw the lines.";
   color.defaultvalue = QVariant(QColor(255, 100, 50, 255));
   color.id = "color";
   configurables.append(color);

   TextureGeneratorSetting lineheight;
   lineheight.name = "Line thickness (%)";
   lineheight.description = "Thickness of each line as a percentage of the texture height.";
   lineheight.defaultvalue = QVariant((int)10);
   lineheight.min = QVariant(0);
   lineheight.max = QVariant(100);
   lineheight.id = "lineheight";
   configurables.append(lineheight);

   TextureGeneratorSetting spacing;
   spacing.name = "Gap size (%)";
   spacing.description = "Gap between adjacent lines as a percentage of the texture height.";
   spacing.defaultvalue = QVariant((int)10);
   spacing.min = QVariant(0);
   spacing.max = QVariant(100);
   spacing.id = "spacing";
   configurables.append(spacing);

   TextureGeneratorSetting offset;
   offset.name = "Pattern offset (%)";
   offset.description = "Shifts the repeating line pattern along its perpendicular axis.";
   offset.defaultvalue = QVariant((int)0);
   offset.min = QVariant(-100);
   offset.max = QVariant(0);
   offset.id = "offset";
   configurables.append(offset);

   TextureGeneratorSetting angle;
   angle.name = "Angle (°)";
   angle.description = "Rotation angle of the lines.";
   angle.defaultvalue = QVariant((double)0);
   angle.min = QVariant(0);
   angle.max = QVariant(180);
   angle.id = "angle";
   configurables.append(angle);
}
void LinesTextureGenerator::generate(QSize size, TexturePixel* destimage,
                                     const QMap<QString, TextureImagePtr>& sourceimages,
                                     const TextureNodeSettings& settings) const {
   if (!destimage || !size.isValid()) {
      return;
   }
   QColor color = settings.value("color").value<QColor>();
   int spacing = settings.value("spacing").toDouble() * size.height() / 100;
   int lineheight = settings.value("lineheight").toDouble() * size.height() / 100;
   int offset = settings.value("offset").toDouble() * size.height() / 100;
   double angle = settings.value("angle").toDouble();
   int period = lineheight + spacing;

   if (sourceimages.contains(QStringLiteral("Input"))) {
      memcpy(destimage, sourceimages.value(QStringLiteral("Input"))->getData(),
             size.width() * size.height() * sizeof(TexturePixel));
   } else {
      memset(destimage, 0, size.width() * size.height() * sizeof(TexturePixel));
   }
   if (period <= 0) {
      return;
   }

   TexturePixel filler(color.red(), color.green(), color.blue(), 255);
   bool invert = false;
   if (angle > 90) {
      invert = true;
      angle = 180 - angle;
   }
   angle = (angle / 180.0) * ((double)M_PI);
   double x1 = 0;
   double y1 = -10000;
   double x1rot = (x1 * cos(angle)) - (y1 * sin(angle));
   double y1rot = (y1 * cos(angle)) + (x1 * sin(angle));
   x1 = x1rot;
   y1 = y1rot;
   double x2 = -x1;
   double y2 = -y1;
   for (int y = 0; y < size.height(); y++) {
      for (int x = 0; x < size.width(); x++) {
         double x3 = x;
         double y3 = y;
         double x4 = -10000;
         double y4 = y;
         double x4rot = ((x4 - x3) * cos(angle)) - ((y4 - y3) * sin(angle)) + x3;
         double y4rot = ((y4 - y3) * cos(angle)) + ((x4 - x3) * sin(angle)) + y3;
         x4 = x4rot;
         y4 = y4rot;
         double intersection_x = ((x2 * y1 - x1 * y2) * (x4 - x) - (x4 * y - x * y4) * (x2 - x1)) /
                                 ((x2 - x1) * (y4 - y) - (x4 - x) * (y2 - y1));
         double intersection_y = ((x2 * y1 - x1 * y2) * (y4 - y) - (x4 * y - x * y4) * (y2 - y1)) /
                                 ((x2 - x1) * (y4 - y) - (x4 - x) * (y2 - y1));
         int distance = sqrt((x - intersection_x) * (x - intersection_x) +
                             (y - intersection_y) * (y - intersection_y));
         if (((distance - offset) % period) > spacing) {
            if (invert) {
               destimage[y * size.width() + size.width() - x - 1] = filler;
            } else {
               destimage[y * size.width() + x] = filler;
            }
         }
      }
   }
}

File: generators/lines.js
const generator = {
  apiVersion: 1,
  name: "Lines",
  description: "Draws evenly spaced parallel lines over an optional input image.",
  type: "generator",
  inputs: ["Canvas"],

  // Each entry creates a control in the node settings panel.
  settings: [
    {
      id: "color",
      type: "color",
      name: "Line colour",
      description: "Colour used to draw the lines.",
      default: { r: 255, g: 100, b: 50, a: 255 },
    },
    {
      id: "lineheight",
      type: "integer",
      name: "Line thickness (%)",
      description: "Thickness of each line relative to the shorter texture dimension.",
      default: 10,
      min: 0,
      max: 100,
    },
    {
      id: "spacing",
      type: "integer",
      name: "Gap size (%)",
      description: "Gap between adjacent lines relative to the shorter texture dimension.",
      default: 10,
      min: 0,
      max: 100,
    },
    {
      id: "offset",
      type: "integer",
      name: "Pattern phase (%)",
      description: "Shifts the pattern through one complete gap-and-line period.",
      default: 0,
      min: 0,
      max: 100,
    },
    {
      id: "angle",
      type: "real",
      name: "Angle (°)",
      description: "Rotation angle of the lines.",
      default: 0,
      min: 0,
      max: 180,
    },
    {
      id: "antialiasing",
      type: "boolean",
      name: "Antialiasing",
      description: "Smooths the edges of the lines.",
      default: true,
    },
  ],

  // This function runs whenever the application renders the node.
  generate(size, settings, output, inputs) {
    // Step 1: copy the optional canvas into the output. The lines will be painted
    // over this image. Without a canvas, start with a transparent image.
    const outputPixels = output.data;
    const canvas = inputs.Canvas?.data;
    if (canvas) outputPixels.set(canvas);
    else outputPixels.fill(0);

    const width = size.width;
    const height = size.height;
    const outputStride = output.stride;

    // Step 2: convert percentages from the settings panel into pixel measurements.
    // Using the shorter dimension prevents a wide or tall texture from stretching
    // the pattern. A 10% line therefore has the same proportions in either shape.
    const referenceSize = Math.min(width, height);
    const lineThickness = settings.lineheight * referenceSize / 100;
    const gapSize = settings.spacing * referenceSize / 100;
    const patternLength = lineThickness + gapSize;

    // A zero-width line has nothing to draw. A zero-length pattern would also make
    // the repeating remainder calculation below impossible.
    if (lineThickness <= 0 || patternLength <= 0 || settings.color.a === 0) return;

    // Phase is a percentage of the complete repeating period. Its full 0–100%
    // range can therefore reach every possible alignment, regardless of the
    // selected line thickness and gap size.
    const patternOffset = settings.offset * patternLength / 100;

    // Step 3: prepare the line angle. JavaScript's sine and cosine functions use
    // radians, while degrees are easier to select in the settings panel.
    const angleInRadians = settings.angle * Math.PI / 180;
    const cosine = Math.cos(angleInRadians);
    const sine = Math.sin(angleInRadians);
    const antialiasing = settings.antialiasing;
    const colour = settings.color;

    // Step 4: visit every output pixel. Projecting (x, y) onto the direction across
    // the lines turns a two-dimensional texture into one repeating measurement.
    for (let y = 0; y < height; ++y) {
      for (let x = 0; x < width; ++x) {
        const distanceAcrossPattern = x * cosine + y * sine - patternOffset;

        // JavaScript keeps the sign of a negative remainder. Adding one complete
        // pattern and taking the remainder again always produces 0..patternLength.
        const positionInPattern = (
          (distanceAcrossPattern % patternLength) + patternLength
        ) % patternLength;
        let coverage;

        if (gapSize <= 0) {
          // No gap means that the line colour covers the complete texture.
          coverage = 1;
        } else if (antialiasing) {
          // A period begins with a gap and ends with a line. Measure the nearest
          // edge, then make the distance negative while the pixel is in the gap.
          // Adding half a pixel turns that signed distance into partial coverage.
          let signedDistance;
          if (positionInPattern >= gapSize) {
            signedDistance = Math.min(
              positionInPattern - gapSize,
              patternLength - positionInPattern,
            );
          } else {
            signedDistance = -Math.min(
              gapSize - positionInPattern,
              positionInPattern,
            );
          }
          coverage = Math.max(0, Math.min(1, signedDistance + 0.5));
        } else {
          // Without antialiasing, a pixel is either completely in a gap or line.
          coverage = positionInPattern >= gapSize ? 1 : 0;
        }

        if (coverage <= 0) continue;

        const outputOffset = y * outputStride + x * 4;

        // Step 5: combine the line colour with the existing canvas pixel. Coverage
        // and the colour's alpha both contribute to the line's effective opacity.
        const lineAlpha = colour.a * coverage / 255;
        const canvasAlpha = outputPixels[outputOffset + 3] / 255;

        // An opaque line replaces the canvas. A transparent canvas has no existing
        // colour to preserve, so it can also be assigned directly.
        if (lineAlpha >= 1 || canvasAlpha === 0) {
          outputPixels[outputOffset] = colour.r;
          outputPixels[outputOffset + 1] = colour.g;
          outputPixels[outputOffset + 2] = colour.b;
          outputPixels[outputOffset + 3] = Math.round(lineAlpha * 255);
        } else {
          // For two translucent colours, retain the visible contribution from both.
          // The output is stored as ordinary, non-premultiplied RGBA.
          const remainingCanvas = 1 - lineAlpha;
          const resultAlpha = lineAlpha + canvasAlpha * remainingCanvas;
          outputPixels[outputOffset] = Math.round(
            (colour.r * lineAlpha
              + outputPixels[outputOffset] * canvasAlpha * remainingCanvas) / resultAlpha,
          );
          outputPixels[outputOffset + 1] = Math.round(
            (colour.g * lineAlpha
              + outputPixels[outputOffset + 1] * canvasAlpha * remainingCanvas) / resultAlpha,
          );
          outputPixels[outputOffset + 2] = Math.round(
            (colour.b * lineAlpha
              + outputPixels[outputOffset + 2] * canvasAlpha * remainingCanvas) / resultAlpha,
          );
          outputPixels[outputOffset + 3] = Math.round(resultAlpha * 255);
        }
      }
    }
  },
};
